### PR TITLE
tools: add component subcommand

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve? 
We need a way to support change the config of other components.

### What is changed and how it works?
This PR adds a component subcommand to manipulate the config manager's configuration.

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -706,7 +740,7 @@ func (s *Server) SetScheduleConfig(cfg config.ScheduleConfig) error {
 			zap.Error(err))
 		return err
 	}
-	log.Info("schedule config is updated", zap.Reflect("new", cfg), zap.Reflect("old", old))
+	// log.Info("schedule config is updated", zap.Reflect("new", cfg), zap.Reflect("old", old))
```
Comment (overvenus) [inline server/server.go]: Uncomment this line?

File: server/api/router.go [hunk 2]
```diff
@@ -188,5 +197,61 @@ func createRouter(prefix string, svr *server.Server) *mux.Router {
...
+					errMsg = reply.(*configpb.UpdateResponse).GetStatus().GetMessage()
+				case "/configpb.Config/Get":
+					errMsg = reply.(*configpb.GetResponse).GetStatus().GetMessage()
+				}
+				if errMsg != "" {
+					return errors.New(errMsg)
+				}
+				return nil
+			}
+			opts := []grpc.DialOption{grpc.WithInsecure(), grpc.WithUnaryInterceptor(UnaryClientInterceptor)}
```
Comment (overvenus) [inline server/api/router.go]: What if the TLS feature is enabled?

**Comment by NingLin-P** [comment]
LGTM

File: server/api/router.go [hunk 3]
```diff
@@ -188,5 +199,77 @@ func createRouter(prefix string, svr *server.Server) *mux.Router {
...
+			componentRouter.Use(newComponentMiddleware(svr).Middleware)
+		}
+		return rootRouter, f
+	}
+	return rootRouter, nil
 }
+
+type customMarshaler struct{}
+
+func (c *customMarshaler) Marshal(v interface{}) ([]byte, error) {
```
Comment (shafreeck) [inline server/api/router.go]: Why are they all empty implementations?

File: tools/pd-ctl/pdctl/command/component_command.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+		Use:   "set [<component>|<componentID>] <option> <value>",
+		Short: "set the option with value",
+		Run:   setComponentConfigCommandFunc,
+	}
+	return sc
+}
+
+func showComponentConfigCommandFunc(cmd *cobra.Command, args []string) {
+	if len(args) != 1 {
+		cmd.Println(cmd.UsageString())
```
Comment (shafreeck) [inline tools/pd-ctl/pdctl/command/component_command.go]: cmd.Usage() is more simpler

File: server/api/middleware.go [hunk 5]
```diff
@@ -56,6 +68,99 @@ type entry struct {
...
+
+func newComponentMiddleware(s *server.Server) componentMiddleware {
+	return componentMiddleware{
+		s:  s,
+		rd: render.New(render.Options{IndentJSON: true}),
+	}
+}
+
+func (m componentMiddleware) Middleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```
Comment (disksing) [inline server/api/middleware.go]: I wonder if it is suitable to use middleware here. After all, the logic here is too complicated.

Comment (disksing) [inline server/api/middleware.go]: I wonder if it is suitable to use middleware here. After all, the logic here is too complicated.

File: server/api/router.go (hunk 3 above)
Comment (rleungx) [reply server/api/router.go]: Because we need to implement Marshaler interface here

File: server/api/middleware.go (hunk 5 above)
Comment (rleungx) [reply server/api/middleware.go]: Any better idea?

**Comment by rleungx** [comment]
PTAL. @disksing

File: server/api/router.go [hunk 6]
```diff
@@ -188,5 +200,81 @@ func createRouter(prefix string, svr *server.Server) *mux.Router {
 	// Deprecated
 	rootRouter.HandleFunc("/ping", func(w http.ResponseWriter, r *http.Request) {}).Methods("GET")
 
-	return rootRouter
+	if svr.GetConfig().EnableDynamicConfig {
+		f := func() {
+			lazyComponentRouter(ctx, svr, apiRouter)
+		}
+		return rootRouter, f
```
Comment (disksing) [inline server/api/router.go]: return rootRouter, func(){lazyComponentRouter(ctx, svr, apiRouter)}

**Pull request merged**